gosax.go

```diff
@@ -161,6 +163,68 @@ func ParseFile(filename string, cb SaxCallbacks) error {
 	return nil
 }
 
+// ParseMem parses XML from in-memory buffer with SAX, applying cb as callbacks.
+func ParseMem(buf bytes.Buffer, cb SaxCallbacks) error {
+
+	// newHandlerStruct zeroes out all the pointers; we assign only those that
+	// are passed as non-nil in SaxCallbacks.
+	SAXhandler := C.newHandlerStruct()
+
+	if cb.StartDocument != nil {
+		SAXhandler.startDocument = C.startDocumentSAXFunc(C.startDocumentCgo)
+	}
+
+	if cb.EndDocument != nil {
+		SAXhandler.endDocument = C.endDocumentSAXFunc(C.endDocumentCgo)
+	}
+
+	if cb.StartElement != nil {
+		SAXhandler.startElement = C.startElementSAXFunc(C.startElementCgo)
+	}
+	// StartElementNoAttr overrides StartElement
+	if cb.StartElementNoAttr != nil {
+		SAXhandler.startElement = C.startElementSAXFunc(C.startElementNoAttrCgo)
+	}
+
+	if cb.EndElement != nil {
+		SAXhandler.endElement = C.endElementSAXFunc(C.endElementCgo)
+	}
+
+	if cb.Characters != nil {
+		SAXhandler.characters = C.charactersSAXFunc(C.charactersCgo)
+	}
+	// CharactersRaw overrides Characters
+	if cb.CharactersRaw != nil {
+		SAXhandler.characters = C.charactersSAXFunc(C.charactersRawCgo)
+	}
+
+	// Pack the callbacks structure into an opaque unsafe.Pointer which we'll
+	// pass to C as user_data, and C will pass it back to our Go callbacks.
+	user_data := pointer.Save(&cb)
+	defer pointer.Unref(user_data)
+
+	// Allocate memory, do not forget to free it after parsing finishes
+	bufPointer := C.malloc(C.size_t(buf.Len()))
+	defer C.free(bufPointer)
+
+	// Create a C buffer from pointer to mem block, and copy raw bytes of Go buffer to it
+	cBuf := (*[1 << 30]byte)(bufPointer)
```

Comment:
Could you replace this with `unsafe.Slice`? It should work starting with Go 1.17

Reply:
my current production version is 1.16 unfortunately

Reply:
Place here a TODO maybe?

Reply:
Would it be a problem for you to use 1.17? Since Go 1.18 is going to be released soon (within a month or so, according to the Go release schedule), using 1.17 seems to be safe in general.

If that's a problem, I don't mind a TODO, though.